### ISSUE TITLE
Not just another .env parser

### ISSUE THREAD

**Issue by jwoertink** [luckyframework/lucky_env#1]
There's a lot of ENV parsers out there already, and the one we [ship with](https://github.com/gdotdesign/cr-dotenv) has been pretty great. I think we can take it up a notch for Lucky integration though... I'm not sure how this would work, but it would be awesome if we had something that looked more like this:

```text
# this is my .env
LUCKY_ENV=development
DEV_PORT(Int32)=3500
ENABLE_CACHE(Bool)=true
```

Then, we had something on the Lucky side that looked like

```crystal
require "lucky_env"

LuckyEnv.development? #=> true
LuckyEnv.lucky_env #=> "development"
LuckyEnv.dev_port #=> 3500
LuckyEnv.enable_cache #=> true
```

By rolling our own, that gives us full control to start adding additional features that other Env parsers may not want.

* We could have auto reflection on `LUCKY_ENV` instead of having that class in your app's config dir.
* A special syntax for declaring non-string types
* Automatic method generation so your app calls these methods instead of having `ENV[""]` laying around. This also plays in to type-safety because how many times have you (I mean "I") done `if ENV["LUCY_ENV"]? == "development"` or something where you misspelled the key, but called `?` because it wasn't really required?

Also, I've found myself following the Lucky pattern we use [here](https://github.com/luckyframework/lucky_cli/blob/master/src/web_app_skeleton/config/server.cr.ecr#L49-L56) quite a bit in my apps. It would be great to know which keys are absolutely required to be loaded in, and if they're not, then we raise a compile-time error to say "You missed one". 

Lastly... I'd like to start thinking about how we get encrypted secrets in to Lucky. This could potentially be that way. Imagine you have this..

```text
# .env.production
LUCKY_ENV=production
AWS_SECRET_KEY=aeihfaowehfgawe
```

Then locally you run a handy new built-in Lucky Cli task `lucky encrypt_env .env.production`
This outputs a file

```text
# env_production.enc
LUCKY_ENV=sfhaoweiubvsadf
AWS_SECRET_KEY=9832y4r29fwe8hw38ehw
```

Then you cann `LuckyEnv.lucky_env #=> "production"` or `LuckyEnv.aws_secret_key #=> "you get it"`.

### Final thoughts

With that said, I could see how parts of this could be done using `method_missing`, and other macro stuff, but there's some talk floating around about `method_missing` maybe not the best thing to use. Also I'm not sure how we load this in at the macro level... Lots to think about

**Comment by ephrimlawrence** [comment]
@jwoertink Is this still on the road map? I want to work on it.

**Comment by jwoertink** [comment]
@ephrimlawrence Love the motivation and wanting to help 🙌  I'm not too sure on this though. At surface level it sounds like a great idea, but I wonder if it could also introduce complications. For example, sometimes your .env has to be shared between a javascript build system like Vite and by Crystal. If we changed the format structure of a standard .env, then you may end up needing to have 2 different `.env` type files that may or may not duplicate values just so Lucky can have this "type-safe" version and Vite/Webpack can use a more standard one....

I did have an idea of an alternate version where maybe your Lucky app just has a config file that is essentially a schema. I've had it happen a few times now where I boot my app and I was actually missing an environment variable which didn't get caught until runtime. Or the opposite case of putting a variable in to my file, then no longer needing it and forgetting to remove it.

```crystal
# config/lucky_env.cr
LuckyEnv.schema do |env|
  # some sort of interface to specify key name, type, and possibly a default???
  env.add("LUCKY_ENV", String)
end
```

I don't know what this looks like or how it works. I'm not opposed to adding something in, but I'm kind of leaning towards we probably shouldn't invent a new `.env` file format just to avoid compatibility issues. I'm open to ideas.

**Comment by ephrimlawrence** [comment]
You're right on the compatibility issues with other build systems. Although I prefer accessing the variables using methods like   `LuckyEnv.aws_secret_key` instead of defining a schema. 

[I have worked on a prototype](https://github.com/ephrimlawrence/lucky_env/tree/typed-env-parser) that uses macros to generate methods for the variables at compile time, based on the proposal.  For compatibility, the default type of a variable is string if not specified. 

Considering that the data types for environment values are limited to string, boolean, integer, and possibly float, we can use regular expressions to determine these types. This avoids explicitly defining types or creating a schema, while still maintaining compatibility with other build systems and ensuring type safety.

**Comment by jwoertink** [comment]
It would be cool to be able to call methods for sure. I wonder if the whole file loading and parsing could just be moved to the macro space, and then if we get the whole hash that could easily just generate the methods fully typed based on String, Int32, Float64, and Bool.

I see you're basically doing that with this `static_load_env`, but it looks like that's using `run()` which makes me sweat a little lol. Since that macro is a known problem in Crystal to cause compilation times to increase, I'm actually looking in to getting rid of every `run()` macro call if I can. I guess I'd have to throw it in my app and see how the compilation time is effected. Right now it takes my app 2+ minutes to compile in development, and nearly 20+ in production.

**Comment by ephrimlawrence** [comment]
I noticed an increase of some milliseconds in the testing time, but a delay of over 2+ minutes is unacceptable. 

It seems that `method_missing` (https://crystal-lang.org/reference/1.16/syntax_and_semantics/macros/hooks.html) only works on a class. Since I’m relatively new to Crystal and not familiar with the advanced features, I might not fully understand this. If I'm not mistaken, `method_missing` may just be syntactic sugar for accessing `ENV['key']`, since the type is only determined at runtime.

**Comment by ephrimlawrence** [comment]
> [@ephrimlawrence](https://github.com/ephrimlawrence) Love the motivation and wanting to help 🙌 I'm not too sure on this though. At surface level it sounds like a great idea, but I wonder if it could also introduce complications. For example, sometimes your .env has to be shared between a javascript build system like Vite and by Crystal. If we changed the format structure of a standard .env, then you may end up needing to have 2 different `.env` type files that may or may not duplicate values just so Lucky can have this "type-safe" version and Vite/Webpack can use a more standard one....
 > 
> I did have an idea of an alternate version where maybe your Lucky app just has a config file that is essentially a schema. I've had it happen a few times now where I boot my app and I was actually missing an environment variable which didn't get caught until runtime. Or the opposite case of putting a variable in to my file, then no longer needing it and forgetting to remove it.
 > 
```
  # config/lucky_env.cr
  LuckyEnv.schema do |env|
    # some sort of interface to specify key name, type, and possibly a default???
    env.add("LUCKY_ENV", String)
  end
```
 > 
 > I don't know what this looks like or how it works. I'm not opposed to adding something in, but I'm kind of leaning towards we probably shouldn't invent a new `.env` file format just to avoid compatibility issues. I'm open to ideas.
> 

Could you provide more details on how you see this working?

**Comment by jwoertink** [comment]
Yeah `method_missing` works on classes and structs, but it only works on instance methods and not class methods which also means LuckyEnv would have to change from being a module.

```crystal
class LuckyEnv
  macro method_missing(call)
    def {{ call }}
      ENV[{{ call.stringify }}]
    end
  end

  def task?
    # ...
  end
end

# this works
LuckyEnv.new.aws_key

# but then this also works when it shouldn't...
LuckyEnv.new.taks?
```

One idea I had was if we used the `read_file()` macro to read and parse the `.env` at compile time, we could then generate the methods with a little bit of guess work.

```crystal
module LuckyEnv
  macro init_env
    {% env_data = read_file(".env") %}
    {%  parsed_data = however_this_happens %}
    {% for key, val in parsed_data %}
      {% type = some_guesswork_on(val) %}
       def self.{{ key }} : {{ type }}
          ENV[{{ key.stringify }}].to_some_conversion.as({{ type }})
        end
     {% end %}
  end
end
```

It's a lot of pseudo code here, but hopefully that gives the idea of what I was thinking. I really don't know if it's possible, or what issues we could run in to, but if it is possible, then this gives us type-safe methods we can call that are already converted to the types we need.

Now, for my previous idea with the schema... again, half-baked idea (because I rarely ever think through an idea all the way through lol), but essentially it's this:

* read schema
* read .env
* compare schema to `ENV` to ensure we have every required ENV var our schema has said is required.

In this option, we're not really getting the type-safety aspect, but instead just a validation tool that says our app requires these ENV vars to exist, and we have an interface that confirms this. Then maybe it throws warnings of "Found X number of unused ENV defined in the .env" or something like that. This becomes helpful to both ensure the app has all of the ENV it needs at boot-time as well as keeping the .env files clean from when you remove some ENV usage.

**Comment by ephrimlawrence** [comment]
Thank you, Jeremy. This is very helpful. I will work with this and see what results come up.

**Comment by ephrimlawrence** [comment]
@jwoertink I have created a PR https://github.com/luckyframework/lucky_env/pull/39. What are your thoughts on it? Can you confirm if the build time has not increased?

**Comment by jwoertink** [comment]
Closed by #39
